bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,190 @@
+            'cost_fn_gen': self.cost,
+        }
+        self.instantiate_options.update(instantiate_options)
+
+    @staticmethod
+    def get_tree_circs(
+        orig_num_cycles: int, circuit_copy: Circuit,
+        cycle_and_ops: list[tuple[int, Operation]],
+    ) -> list[Circuit]:
+        # Implement recursively for now, if slow then fix
+        all_circs = [circuit_copy.copy()]
+        for cycle, op in cycle_and_ops:
+            new_circs = []
+            for circ in all_circs:
+                idx_shift = orig_num_cycles - circ.num_cycles
+                new_cycle = cycle - idx_shift
+                work_copy = circ.copy()
+                work_copy.pop((new_cycle, op.location[0]))
+                new_circs.append(work_copy)
+                new_circs.append(circ)
+
+            all_circs = new_circs
+
+        all_circs = sorted(all_circs, key=lambda x: x.num_operations)
+
+        return all_circs
+
+    async def run(self, circuit: Circuit, data: PassData) -> None:
+        """Perform the pass's operation, see :class:`BasePass` for more."""
+        instantiate_options = self.instantiate_options.copy()
+        if 'seed' not in instantiate_options:
+            instantiate_options['seed'] = data.seed
+
+        start = 'left' if self.start_from_left else 'right'
+        _logger.debug(f'Starting scanning gate removal on the {start}.')
+
+        target = self.get_target(circuit, data)
+        # target = None
+
+        circuit_copy = circuit.copy()
+        reverse_iter = not self.start_from_left
+
+        ops_left = list(circuit.operations_with_cycles(reverse=reverse_iter))
+        print(
+            f'Starting Scan with tree depth {self.tree_depth}'
+            ' on circuit with {len(ops_left)} gates',
+        )
+
+        while ops_left:
+            chunk = ops_left[:self.tree_depth]
+            ops_left = ops_left[self.tree_depth:]
+
+            # Circuits of size 2 ** tree_depth - 1,
+            # ranked in order of most to fewest deletions
+            all_circs = TreeScanningGateRemovalPass.get_tree_circs(
+                circuit.num_cycles, circuit_copy, chunk,
+            )
+            # Remove circuit with no gates deleted
+            all_circs = all_circs[:-1]
+
+            _logger.debug(
+                'Attempting removal of operation of'
+                f' {self.tree_depth} operations.',
+            )
+
+            instantiated_circuits = await get_runtime().map(
+                Circuit.instantiate,
+                all_circs,
+                target=target,
+                **instantiate_options,
+            )
+
+            dists = [self.cost(c, target) for c in instantiated_circuits]
+
+            # Pick least count with least dist
+            for i, dist in enumerate(dists):
+                if dist < self.success_threshold:
+                    _logger.debug(
+                        f'Successfully switched to circuit {i}'
+                        ' of {2 ** self.tree_depth}.',
```

Comment:
From a logging perspective, this is not helpful. We should instead say something like... "Successfully removed X CNOTs Y U3s" or something like that. Plus f-string issue.

Consider a situation where a user is running on Perlmutter for 12 hours and the time-limit hits. You go to the logs to try and gather some amount of information about how it went to see if its worth restarting or if it did well. You look at the logs and you see a bunch of "Successfully switched to circuit X of of {2 ** Y}." That may not be helpful. Instead if I can count how many successfully removed CNOTs, then I can gather that in 12 hours I made Z amount of progress or something. I know this means that you will have to do more than a simple one-line log call to gather this information, but it is worth doing good logging.